ydb/_topic_writer/topic_writer_asyncio.py

```diff
@@ -350,6 +402,124 @@ async def _connection_loop(self):
                         task.cancel()
                     await asyncio.wait(pending)
 
+    async def _encode_loop(self):
+        while True:
+            messages = await self._messages_for_encode.get()
+            while not self._messages_for_encode.empty():
+                messages.extend(self._messages_for_encode.get_nowait())
+
+            batch_codec = await self._codec_selector(messages)
+            await self._encode_data_inplace(batch_codec, messages)
+            self._add_messages_to_send_queue(messages)
+
+    async def _encode_data_inplace(
+        self, codec: PublicCodec, messages: List[InternalMessage]
+    ):
+        if codec == PublicCodec.RAW:
+            return
+
+        eventloop = asyncio.get_running_loop()
+        encode_waiters = []
+        encoder_function = self._codec_functions[codec]
+
+        for message in messages:
+            encoded_data_futures = eventloop.run_in_executor(
+                self._encode_executor, encoder_function, message.get_bytes()
+            )
+            encode_waiters.append(encoded_data_futures)
+
+        encoded_datas = await asyncio.gather(*encode_waiters)
+
+        for index, data in enumerate(encoded_datas):
+            message = messages[index]
+            message.codec = codec
+            message.data = data
+
+    async def _codec_selector(self, messages: List[InternalMessage]) -> PublicCodec:
+        if self._codec is not None:
+            return self._codec
+
+        if self._codec_selector_last_codec is None:
+            available_codecs = await self._get_available_codecs()
+
+            # use every of available encoders at start for prevent problems
+            # with rare used encoders (on writer or reader side)
+            if self._codec_selector_batch_num < len(available_codecs):
+                codec_index = self._codec_selector_batch_num % len(available_codecs)
+                codec = available_codecs[codec_index]
+            else:
+                codec = await self._codec_selector_by_check_compress(messages)
+                self._codec_selector_last_codec = codec
+        else:
+            if (
+                self._codec_selector_batch_num
+                % self._codec_selector_check_batches_interval
+                == 0
+            ):
+                self._codec_selector_last_codec = (
+                    await self._codec_selector_by_check_compress(messages)
+                )
+            codec = self._codec_selector_last_codec
+        self._codec_selector_batch_num += 1
+        return codec
+
+    async def _get_available_codecs(self) -> List[PublicCodec]:
```

Comment:
Get allowed codec from server and intersect it with supported codecs with encoders map.

If server side has no codec restriction - use raw and gzip only (for better compatibility with readers).